Bug: Return Uni<> and enable structuredContent not working

Hey I'm not sure if this is intended or just a bug but when using `@Tool` with a `Uni<>` return type and enabling structuredContent, the project fails to build. 
When `structuredContent=false` everything works fine.

```Java
    @Tool(description = "Fetch connector data from Connectors API", structuredContent = true)
    public Uni<JsonObject> connectors(){
        return connectorsService.getOrFetchDataAsnyc();
    }
```

```bash
2025-10-31 10:50:02,265 ERROR [io.qua.run.Application] (Quarkus Main Thread) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:135)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NoSuchMethodError: 'java.util.function.Function io.quarkiverse.mcp.server.runtime.ToolStructuredContentResultMapper.uni()'
        at io.quarkiverse.mcp.server.runtime.McpMetadata_Impl.meta$2(Unknown Source)
        at io.quarkiverse.mcp.server.runtime.McpMetadata_Impl.tools(Unknown Source)
        at io.quarkiverse.mcp.server.runtime.ToolManagerImpl.<init>(ToolManagerImpl.java:71)
        at io.quarkiverse.mcp.server.runtime.ToolManagerImpl_Bean.doCreate(Unknown Source)
        at io.quarkiverse.mcp.server.runtime.ToolManagerImpl_Bean.create(Unknown Source)
        at io.quarkiverse.mcp.server.runtime.ToolManagerImpl_Bean.create(Unknown Source)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
        at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:32)
        at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
        at io.quarkus.arc.impl.ComputingCacheContextInstances.computeIfAbsent(ComputingCacheContextInstances.java:19)
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
        at io.quarkiverse.mcp.server.runtime.ToolManagerImpl_Bean.get(Unknown Source)
        at io.quarkiverse.mcp.server.runtime.ToolManagerImpl_Bean.get(Unknown Source)
        at io.quarkiverse.mcp.server.sse.runtime.StreamableHttpMcpMessageHandler_Bean.doCreate(Unknown Source)
        at io.quarkiverse.mcp.server.sse.runtime.StreamableHttpMcpMessageHandler_Bean.create(Unknown Source)
        at io.quarkiverse.mcp.server.sse.runtime.StreamableHttpMcpMessageHandler_Bean.create(Unknown Source)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
        at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:32)
        at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
        at io.quarkus.arc.impl.ComputingCacheContextInstances.computeIfAbsent(ComputingCacheContextInstances.java:19)
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
        at io.quarkiverse.mcp.server.sse.runtime.StreamableHttpMcpMessageHandler_Bean.get(Unknown Source)
        at io.quarkiverse.mcp.server.sse.runtime.StreamableHttpMcpMessageHandler_Bean.get(Unknown Source)
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:570)
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:550)
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:583)
        at io.quarkus.arc.impl.ArcContainerImpl.instanceHandle(ArcContainerImpl.java:545)
        at io.quarkus.arc.impl.ArcContainerImpl.instance(ArcContainerImpl.java:300)
        at io.quarkiverse.mcp.server.sse.runtime.SseMcpServerRecorder.createMcpEndpointHandler(SseMcpServerRecorder.java:62)
        at io.quarkus.runner.recorded.SseMcpServerProcessor$registerEndpoints2072145912.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.SseMcpServerProcessor$registerEndpoints2072145912.deploy(Unknown Source)
```

--- a/core/deployment/src/main/java/io/quarkiverse/mcp/server/deployment/McpServerProcessor.java
+++ b/core/deployment/src/main/java/io/quarkiverse/mcp/server/deployment/McpServerProcessor.java
@@ -344,8 +344,11 @@ class McpServerProcessor {
                         AnnotationValue outputSchemaValue = featureAnnotation.value("outputSchema");
                         if (outputSchemaValue != null) {
                             AnnotationValue outputSchemaFromValue = outputSchemaValue.asNested().value("from");
-                            if (outputSchemaFromValue != null)
+                            if (outputSchemaFromValue != null) {
                                 outputSchemaFrom = outputSchemaFromValue.asClass();
+                            } else {
+                                outputSchemaFrom = outputSchemaFromReturnType(method.returnType());
+                            }
                             AnnotationValue outputSchemaGeneratorValue = outputSchemaValue.asNested().value("generator");
                             if (outputSchemaGeneratorValue != null) {
                                 outputSchemaGenerator = outputSchemaGeneratorValue.asClass();
@@ -353,7 +356,7 @@ class McpServerProcessor {
                                 outputSchemaGenerator = ClassType.create(GlobalOutputSchemaGenerator.class);
                             }
                         } else if (structuredContent) {
-                            outputSchemaFrom = ClassType.create(method.returnType().name());
+                            outputSchemaFrom = outputSchemaFromReturnType(method.returnType());
                             outputSchemaGenerator = ClassType.create(GlobalOutputSchemaGenerator.class);
                         }
 
@@ -457,6 +460,13 @@ class McpServerProcessor {
                 }
             }
         }
+    }
+
+    private ClassType outputSchemaFromReturnType(org.jboss.jandex.Type returnType) {
+        if (returnType.name().equals(DotNames.UNI)) {
+            return ClassType.create(returnType.asParameterizedType().arguments().get(0).name());
+        }
+        return ClassType.create(returnType.name());
     }
 
     private Content.Annotations parseResourceAnnotations(AnnotationInstance featureAnnotation) {
@@ -1282,21 +1292,24 @@ class McpServerProcessor {
                 MethodDescriptor.ofMethod(InstanceHandle.class, "get", Object.class),
                 instance);
         if (DotNames.UNI.equals(returnType.name())) {
-            if (featureMethod.getFeature() != PROMPT
-                    && DotNames.LIST.equals(returnType.asParameterizedType().arguments().get(0).name())) {
+            if (useUniList(featureMethod, returnType.asParameterizedType().arguments().get(0))) {
                 mapper = bytecode.invokeVirtualMethod(
                         MethodDescriptor.ofMethod(mapperClazz, "uniList", Function.class), mapper);
             } else {
                 mapper = bytecode.invokeVirtualMethod(
                         MethodDescriptor.ofMethod(mapperClazz, "uni", Function.class), mapper);
             }
-        } else if (featureMethod.getFeature() != PROMPT
-                && !featureMethod.isStructuredContent()
-                && DotNames.LIST.equals(returnType.name())) {
+        } else if (useUniList(featureMethod, returnType)) {
             mapper = bytecode.invokeVirtualMethod(
                     MethodDescriptor.ofMethod(mapperClazz, "list", Function.class), mapper);
         }
         return mapper;
+    }
+
+    private boolean useUniList(FeatureMethodBuildItem featureMethod, org.jboss.jandex.Type type) {
+        return featureMethod.getFeature() != PROMPT
+                && !featureMethod.isStructuredContent()
+                && DotNames.LIST.equals(type.name());
     }
 
     ResultHandle toolResultMapper(FeatureMethodBuildItem featureMethod, BytecodeCreator bytecode,

--- a/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ToolManagerImpl.java
+++ b/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ToolManagerImpl.java
@@ -230,11 +230,13 @@ public class ToolManagerImpl extends FeatureManagerBase<ToolResponse, ToolInfo> 
             Class<? extends OutputSchemaGenerator> outputSchemaGeneratorClass = metadata.info().outputSchemaGenerator();
             if (outputSchemaGeneratorClass != null) {
                 Object outputSchema;
+                Class<?> outputSchemaFrom = metadata.info().outputSchemaFrom();
                 if (GlobalOutputSchemaGenerator.class.equals(outputSchemaGeneratorClass)) {
-                    outputSchema = globalOutputSchemaGenerator.generate(metadata.info().outputSchemaFrom());
+                    outputSchema = globalOutputSchemaGenerator.generate(outputSchemaFrom);
                 } else {
-                    outputSchema = outputSchemaGenerator.select(outputSchemaGeneratorClass).get()
-                            .generate(metadata.info().outputSchemaFrom());
+                    outputSchema = outputSchemaGenerator.select(outputSchemaGeneratorClass)
+                            .get()
+                            .generate(outputSchemaFrom);
                 }
                 tool.put("outputSchema", outputSchema);
             }

--- a/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ToolMessageHandler.java
+++ b/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ToolMessageHandler.java
@@ -49,7 +49,14 @@ class ToolMessageHandler extends MessageHandler {
         JsonObject result = new JsonObject().put("tools", tools);
         Page<ToolManager.ToolInfo> page = manager.fetchPage(mcpRequest, cursor, pageSize);
         for (ToolManager.ToolInfo info : page) {
-            tools.add(info.asJson());
+            try {
+                tools.add(info.asJson());
+            } catch (McpException e) {
+                return mcpRequest.sender().sendError(id, e.getJsonRpcErrorCode(), e.getMessage());
+            } catch (Exception e) {
+                LOG.errorf(e, "Unable to encode Tool [%s] as JSON", info.name());
+                return mcpRequest.sender().sendInternalError(id);
+            }
         }
         if (page.hasNextCursor()) {
             ToolManager.ToolInfo last = page.lastInfo();

--- a/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ToolStructuredContentResultMapper.java
+++ b/core/runtime/src/main/java/io/quarkiverse/mcp/server/runtime/ToolStructuredContentResultMapper.java
@@ -39,6 +39,10 @@ public class ToolStructuredContentResultMapper implements Function<Result<Object
         };
     }
 
+    public Function<Result<Uni<Object>>, Uni<ToolResponse>> uni() {
+        return uni;
+    }
+
     @Override
     public Uni<ToolResponse> apply(Result<Object> r) {
         boolean compatMode = true;

--- a/transports/sse/deployment/src/test/java/io/quarkiverse/mcp/server/test/tools/structcontent/FailingOutputSchemaGeneratorTest.java
+++ b/transports/sse/deployment/src/test/java/io/quarkiverse/mcp/server/test/tools/structcontent/FailingOutputSchemaGeneratorTest.java
@@ -1,0 +1,85 @@
+package io.quarkiverse.mcp.server.test.tools.structcontent;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import jakarta.inject.Singleton;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkiverse.mcp.server.JsonRpcErrorCodes;
+import io.quarkiverse.mcp.server.OutputSchemaGenerator;
+import io.quarkiverse.mcp.server.Tool;
+import io.quarkiverse.mcp.server.Tool.OutputSchema;
+import io.quarkiverse.mcp.server.test.McpAssured;
+import io.quarkiverse.mcp.server.test.McpAssured.McpStreamableTestClient;
+import io.quarkiverse.mcp.server.test.McpServerTest;
+import io.quarkus.test.QuarkusUnitTest;
+import io.vertx.core.json.JsonObject;
+
+public class FailingOutputSchemaGeneratorTest extends McpServerTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = defaultConfig(5000)
+            .withApplicationRoot(
+                    root -> root.addClasses(MyTools.class, MyPojo.class, FailingSchemaGenerator.class));
+
+    @Test
+    public void testContents() {
+        McpStreamableTestClient client = McpAssured.newConnectedStreamableClient();
+
+        client.when()
+                .toolsList()
+                .withErrorAssert(error -> {
+                    assertEquals(JsonRpcErrorCodes.INTERNAL_ERROR, error.code());
+                })
+                .send()
+                .toolsCall("delta", toolResponse -> {
+                    assertEquals(0, toolResponse.content().size());
+                    assertNotNull(toolResponse.structuredContent());
+                    if (toolResponse.structuredContent() instanceof JsonObject json) {
+                        assertEquals(10, json.getInteger("val"));
+                    } else {
+                        fail("Not a JsonObject");
+                    }
+                })
+                .thenAssertResults();
+    }
+
+    public static class MyPojo {
+
+        private int val;
+
+        public int getVal() {
+            return val;
+        }
+
+        public void setVal(int val) {
+            this.val = val;
+        }
+
+    }
+
+    public static class MyTools {
+
+        @Tool(description = "With custom schema generator", structuredContent = true, outputSchema = @OutputSchema(generator = FailingSchemaGenerator.class))
+        MyPojo delta() {
+            MyPojo pojo = new MyPojo();
+            pojo.setVal(10);
+            return pojo;
+        }
+    }
+
+    @Singleton
+    public static class FailingSchemaGenerator implements OutputSchemaGenerator {
+
+        @Override
+        public Object generate(Class<?> from) {
+            throw new IllegalStateException("Bum!");
+        }
+
+    }
+
+}

--- a/transports/sse/deployment/src/test/java/io/quarkiverse/mcp/server/test/tools/structcontent/ToolStructuredContentTest.java
+++ b/transports/sse/deployment/src/test/java/io/quarkiverse/mcp/server/test/tools/structcontent/ToolStructuredContentTest.java
@@ -4,6 +4,8 @@ import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.fail;
 
+import java.util.Objects;
+
 import jakarta.inject.Inject;
 import jakarta.inject.Singleton;
 
@@ -20,6 +22,7 @@ import io.quarkiverse.mcp.server.test.McpAssured.McpStreamableTestClient;
 import io.quarkiverse.mcp.server.test.McpServerTest;
 import io.quarkus.runtime.Startup;
 import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.mutiny.Uni;
 import io.vertx.core.json.JsonObject;
 
 public class ToolStructuredContentTest extends McpServerTest {
@@ -35,7 +38,7 @@ public class ToolStructuredContentTest extends McpServerTest {
 
         client.when()
                 .toolsList(page -> {
-                    assertEquals(4, page.tools().size());
+                    assertEquals(5, page.tools().size());
                     JsonObject alphaSchema = page.findByName("alpha").outputSchema();
                     assertNotNull(alphaSchema);
                     assertEquals("integer", alphaSchema.getJsonObject("properties").getJsonObject("val").getString("type"));
@@ -49,6 +52,9 @@ public class ToolStructuredContentTest extends McpServerTest {
                     assertNotNull(deltaSchema);
                     assertEquals("integer", deltaSchema.getJsonObject("properties").getJsonObject("val").getString("type"));
                     assertEquals("baz", deltaSchema.getJsonObject("properties").getJsonObject("val").getString("foo"));
+                    JsonObject echoSchema = page.findByName("echo").outputSchema();
+                    assertNotNull(echoSchema);
+                    assertEquals("integer", echoSchema.getJsonObject("properties").getJsonObject("val").getString("type"));
                 })
                 .toolsCall("alpha", toolResponse -> {
                     assertEquals(0, toolResponse.content().size());
@@ -82,6 +88,15 @@ public class ToolStructuredContentTest extends McpServerTest {
                     assertNotNull(toolResponse.structuredContent());
                     if (toolResponse.structuredContent() instanceof JsonObject json) {
                         assertEquals(10, json.getInteger("val"));
+                    } else {
+                        fail("Not a JsonObject");
+                    }
+                })
+                .toolsCall("echo", toolResponse -> {
+                    assertEquals(0, toolResponse.content().size());
+                    assertNotNull(toolResponse.structuredContent());
+                    if (toolResponse.structuredContent() instanceof JsonObject json) {
+                        assertEquals(7, json.getInteger("val"));
                     } else {
                         fail("Not a JsonObject");
                     }
@@ -142,6 +157,13 @@ public class ToolStructuredContentTest extends McpServerTest {
                     .register();
         }
 
+        @Tool(description = "Use echo!", structuredContent = true)
+        Uni<MyPojo> echo() {
+            MyPojo pojo = new MyPojo();
+            pojo.setVal(7);
+            return Uni.createFrom().item(pojo);
+        }
+
     }
 
     @Singleton
@@ -149,6 +171,7 @@ public class ToolStructuredContentTest extends McpServerTest {
 
         @Override
         public Object generate(Class<?> from) {
+            Objects.requireNonNull(from);
             // "type" : "object",
             // "properties" : {
             //    "val" : {

